File: HW_2_Diffie_Hellman/Diffie-Hellman/Diffie-Hellman/utils/RotationUtils.cs
using System;
using System.Linq;

namespace Diffie_Hellman.utils
{
    public static class RotationUtils
    {
        public const int MaxUtf32 = 0x10FFFF;
        public enum StringType
        {
            PlainText,
            CipherText,
            KeyText
        }
        public static int NormalizeCharValue(int charValue, StringType type)
        {
            // Normalize the actual number we're going to shift with
            // In c#, number field springs up from number 0 and flows outwards
            // Hence, modulo operator works weirdly.
            // Assume |A| > |B| and A > 0 and B < 0
            // Modulo operator, in B % A should return ||B| - |A||
            // But what it actually returns, is -||B|%|A||
            // Next line remedies that
            while (charValue < 0)
            {
                charValue = MaxUtf32 + charValue;
            }
            // loop back on the number field up from 0
            charValue %= MaxUtf32;
            // If character value lands in the surrogate number range, get out of there
            // If encrypting, make it 0xe000, otherwise 0xd799 (right before and after the range)
            if (charValue is >= 0xd800 and <= 0xdfff)
            {
                charValue = type == StringType.CipherText ? 0xd800 - 1 : 0xdfff + 1 ;
            }

            return charValue;
        }

        public static int RotateCharacterValue(int charValue, int keyValue, StringType type)
        {
            if(type == StringType.KeyText) throw new Exception($"Illegal argument value: {type}");
            // Rotate character value. If encrypting, add key value, otherwise subtract
            charValue = type == StringType.CipherText ? charValue - keyValue : charValue + keyValue;
            charValue = NormalizeCharValue(charValue, type);
            while (char.IsControl(char.ConvertFromUtf32(charValue), 0))
            {
                // Rotate character value out of the undesirable (control character) number range.
                // If encrypting, add key value, otherwise subtract
                charValue = type == StringType.CipherText ? charValue - 1 : charValue + 1;
                charValue = NormalizeCharValue(charValue, type);
            }

            return charValue;
        }

        

        public static string DoCaesar(string shiftable, int shiftInt, StringType type )
        {
            // Calculate the actual number we're going to shift with
            // In c#, number field springs up from number 0 and flows outwards
            // Hence, modulo operator works weirdly.
            // Assume |A| > |B| and A > 0 and B < 0
            // Modulo operator, in B % A should return ||B| - |A||
            // But what it actually returns, is -||B|%|A||
            // Next line remedies that
            while (shiftInt < 0)
            {
                shiftInt = MaxUtf32 + shiftInt;
            }
            // regular modulo operation
            shiftInt = shiftInt % MaxUtf32;
            Console.WriteLine($"Shift: {shiftInt}");
            var ciphertext = "";
            // rotate each character by the given shift
            foreach (var unicodeCodePoint in shiftable.GetUnicodeCodePoints())
            {
                var unicodeValue = RotateCharacterValue(unicodeCodePoint, shiftInt, type);
                ciphertext += char.ConvertFromUtf32(unicodeValue);
            }

            return ciphertext;
        }

        public static string DoVigenere(string inputText, string keyString, StringType type)
        {
            // Normalizing the key string
            if (keyString.GetUtf32Length() != inputText.GetUtf32Length())
            {
                //Make sure that key string is at least the size of cipher string, or more
                while (keyString.GetUtf32Length() < inputText.GetUtf32Length())
                {
                    keyString += keyString;
                }
                //Make sure they're the same length by truncating the possibly longer key string
                keyString = keyString.Utf32Substring(0, inputText.GetUtf32Length());
            }
            var cipherText = "";
            // rotate the value of ith character in plaintext by the UTF32 value of the ith character in 
            // key string
            foreach (var codePoints in inputText.GetUnicodeCodePoints().Zip(keyString.GetUnicodeCodePoints()))
            {
                var unicodeValue = RotateCharacterValue(codePoints.First, codePoints.Second, type);
                cipherText += char.ConvertFromUtf32(unicodeValue);
            }

            return cipherText;
        }
    }
}

File: src/rotationutils/rotationutils.go
package rotationutils

import (
	"encoding/base64"
	"encoding/binary"
	"unicode"
	"unicode/utf8"

	"github.com/m0rdhau/Web_App_Sec/src/cryptoutils"
)

type StringType int

const (
	PlainText StringType = iota
	CipherText
	KeyText
	RSA
)

const MaxUTF int32 = 0x10FFFF

// Normalize the actual number we're going to shift with
// In go, number field springs up from number 0 and flows outwards
// Hence, modulo operator works weirdly.
// Assume |A| > |B| and A > 0 and B < 0
// Modulo operator, in B % A should return ||B| - |A||
// But what it actually returns, is -||B|%|A||
// This function remedies that
func NormalizeCharValue(charValue *int32, strtype StringType, isShift bool) {
	for *charValue < 0 {
		*charValue = MaxUTF + *charValue
	}
	// loop back on the number field up from 0
	*charValue %= MaxUTF
	// If character value lands in the surrogate number range, get out of there
	// If encrypting, make it 0xe000, otherwise 0xd799 (right before and after the range)
	if !isShift {
		if *charValue >= 0xd800 && *charValue <= 0xdfff {
			if strtype == CipherText {
				*charValue = 0xd800 - 1
			} else {
				*charValue = 0xdfff + 1
			}
		}
	}
}

// Rotate character value. If encrypting, add key value, otherwise subtract
// the function also makes sure that the rune we return is not a control character
func RotateCharacterValue(char rune, key rune, strtype StringType) rune {
	charValue := int32(char)
	keyValue := int32(key)
	//this should not happen
	if strtype == KeyText {
		panic("Illegal string type")
	}
	if strtype == CipherText {
		charValue -= keyValue
	} else {
		charValue += keyValue
	}
	NormalizeCharValue(&charValue, strtype, false)
	for unicode.IsControl(rune(charValue)) {
		if strtype == CipherText {
			charValue--
		} else {
			charValue++
		}
		NormalizeCharValue(&charValue, strtype, false)
	}

	return rune(charValue)
}

func DoCaesar(shiftable string, shiftint int32, strtype StringType) (string, int32) {
	//normalize to calculate what the shift is. We'll have to later return it
	NormalizeCharValue(&shiftint, CipherText, true)
	CipherText := ""
	for _, char := range shiftable {
		CipherText += string(RotateCharacterValue(char, rune(shiftint), strtype))
	}
	return CipherText, shiftint
}

func DoVigenere(inputText string, keyString string, strtype StringType) string {
	// normalize the key so that it's the same size as input
	// at first, make sure it's at least the same or bigger
	if utf8.RuneCountInString(inputText) != utf8.RuneCountInString(keyString) {
		for utf8.RuneCountInString(keyString) < utf8.RuneCountInString(inputText) {
			keyString += keyString
		}
	}
	CipherText := ""
	inputRunes := []rune(inputText)
	// then cut it down to size
	keyRunes := []rune(keyString)[0:len(inputRunes)]
	for i := range inputRunes {
		CipherText += string(RotateCharacterValue(inputRunes[i], keyRunes[i], strtype))
	}
	return CipherText
}

func DoRSA(inputText string, mod uint64, exp uint64, strtype StringType) (string, error) {
	CipherText := ""
	if strtype == PlainText {
		inputRunes := []rune(inputText)
		inputEncrypted := make([]byte, 0)
		for i := range inputRunes {
			encRuneBytes := make([]byte, 8)
			encrypted := cryptoutils.Modpow(mod, exp, uint64(inputRunes[i]))
			binary.LittleEndian.PutUint64(encRuneBytes, encrypted)
			inputEncrypted = append(inputEncrypted, encRuneBytes...)
		}
		b64Encoded := base64.StdEncoding.EncodeToString(inputEncrypted)
		return b64Encoded, nil
	} else {
		b64Bytes, err := base64.StdEncoding.DecodeString(inputText)
		if err != nil {
			return "", nil
		}
		outputInts := make([]rune, 0)
		for i := 0; i < len(b64Bytes); i += 8 {
			currBytes := b64Bytes[i : i+8]
			currInt := binary.LittleEndian.Uint64(currBytes)
			outputInts = append(outputInts, rune(cryptoutils.Modpow(mod, exp, currInt)))
		}
		CipherText = string(outputInts)
	}
	return CipherText, nil
}
